### PR TITLE
Optimize the private getMode_ function for production build

### PR DESCRIPTION
**Pull request by erwinmombay**
Closes https://github.com/ampproject/amphtml/issues/4086

### CODE REVIEW

**Comment by erwinmombay** [comment]
screen shot 2016-07-17 at 10 06 39 pm

**Comment by erwinmombay** [comment]
@cramforce PTAL

**Comment by cramforce** [comment]
Wow. That is quite the large hammer for fixing this issue. I'm wondering if, even if we don't get defines to do the job for us, could just emulate a type of constant replacement instead?

Like

`const OVERRIDE_NOT_LOCALDEV = true`

And then replace that?

**Comment by erwinmombay** [comment]
@cramforce i've unfortunately been having a hard time using `--define` overrides (i'm sure i'm just not using it right) as i one of my other PR to silence the warnings, i couldn't get it to work at all once i had 2 modules wherein 1 was importing the other.

**Comment by cramforce** [comment]
K, but how about a simple `const` value replacement to let the code short circuit?

**Comment by erwinmombay** [comment]
@cramforce yeah that will definitely work, will update

**Comment by erwinmombay** [comment]
@cramforce PTAL

File: src/mode.js [hunk 1]
```diff
@@ -65,11 +65,14 @@ export function setModeForTesting(m) {
  * @return {!ModeDef}
  */
 function getMode_() {
+  const OVERRIDE_IS_NOT_PROD = true;
```
Comment (jridgewell) [inline src/mode.js]: Can we call this `IS_DEV`?

Comment (cramforce) [inline src/mode.js]: Or `OVERRIDE_IS_DEV`

File: build-system/runner/src/org/ampproject/AmpPass.java [hunk 2]
```diff
@@ -74,9 +74,21 @@ public AmpPass(AbstractCompiler compiler, boolean isProd, Set<String> stripTypeS
     } else if (isProd && isFunctionInvokeAndPropAccess(n, "$mode.getMode",
         ImmutableSet.of("minified"))) {
       replaceWithBooleanExpression(true, n, parent);
+    } else if (isProd && isLValueName(n, "OVERRIDE_IS_NOT_PROD")) {
```
Comment (cramforce) [inline build-system/runner/src/org/ampproject/AmpPass.java]: Pass in replacement map?

File: src/mode.js (hunk 1 above)
Comment (jridgewell) [inline src/mode.js]: `ANYTHING_THAT_DOESN"T_HAVE_A_NEGATION_IN_THE_NAME` 😁

Comment (erwinmombay) [inline src/mode.js]: haha i intentionally made the name non intuitive to not collide with anything in the future lol, will fix

File: build-system/runner/src/org/ampproject/AmpPass.java (hunk 2 above)
Comment (erwinmombay) [inline build-system/runner/src/org/ampproject/AmpPass.java]: done. added immutablemap coming from the runner passing it into the custom pass.

File: src/mode.js (hunk 1 above)
Comment (erwinmombay) [inline src/mode.js]: done. switched to `IS_DEV` and added presubmit check for it

**Comment by erwinmombay** [comment]
@cramforce PTAL

**Comment by erwinmombay** [comment]
@cramforce PTAL

**Comment by cramforce** [comment]
LGTM

**Pull request merged**